sheet08/sheet08solutions.ipynb

```diff
@@ -327,373 +264,223 @@
-    "        self.epsilon = epsilon\n",
+    "        self.layers = layers\n",
     "\n",
-    "    def activation(self, X):\n",
-    "        \"\"\"\n",
-    "        The activation function. Prepends a 1 to X for the\n",
-    "        bias and calculates the activation function of the \n",
-    "        perceptron.\n",
+    "    def activate(self, x):\n",
+    "        \"\"\"Activate network and return the last layer's output.\n",
     "\n",
     "        Args:\n",
-    "            X   the data point, should be a numpy\n",
-    "                array or a 1xN numpy matrix\n",
+    "            x: Vector of input values.\n",
+    "\n",
     "        Returns:\n",
-    "            1   if the activation of X is bigger than 0\n",
-    "            0   else\n",
+    "            Vector of output values from the last layer of the network\n",
+    "            after propagating forward through the network.\n",
     "        \"\"\"\n",
-    "        return 1 if np.append(1, X) @ self.w > 0 else 0\n",
+    "        for layer in self.layers:\n",
+    "            x = layer.activate(x)\n",
+    "        return x\n",
     "\n",
-    "    def adaption(self, X, t):\n",
-    "        \"\"\"\n",
-    "        Trains the perceptron. Adjusts the weights according to \n",
-    "        the learning rate and the error between the activation and the \n",
-    "        label (delta).\n",
+    "    def adapt(self, x, t, rate=0.03):\n",
+    "        \"\"\"Adapt the whole network given an input and expected output.\n",
     "\n",
     "        Args:\n",
-    "            X   the data point, should be a numpy\n",
-    "                array or a 1xN numpy matrix\n",
-    "            t   the label\n",
+    "            x: Vector of input values.\n",
+    "            t: Vector of target values (expected outputs).\n",
+    "            rate: Learning rate.\n",
     "        \"\"\"\n",
-    "        self.w += self.epsilon * (t - self.activation(X)) * np.append(1, X)\n",
+    "        # Activate each layer and collect intermediate outputs.\n",
+    "        # outputs = []  # TODO\n",
+    "        outputs = [x]\n",
+    "        for layer in self.layers:\n",
+    "            outputs.append(layer.activate(outputs[-1]))\n",
+    "\n",
+    "        # Calculate error between t and network output.\n",
+    "        # e = np.array([])  # TODO\n",
+    "        e = t - outputs[-1]\n",
     "\n",
-    "    def __repr__(self):\n",
-    "        return '{}({}, {})'.format(self.__class__.__name__, len(self.w) - 1, self.epsilon)"
+    "        # Backpropagate error through the network computing\n",
+    "        # intermediate delta and adapting each layer.\n",
+    "        for k in range(len(self.layers) - 1, 0, -1):\n",
+    "            delta = (outputs[k + 1] * (1 - outputs[k + 1])) * e\n",
+    "            weights = self.layers[k].get_weights_matrix()\n",
+    "            e = (weights @ delta)[1:]\n",
+    "            self.layers[k].adapt(outputs[k], delta, rate)"
```

Comment:
Still does not tackle the problem with the weird `k` and `k+1` when comparing to the formula. Its basically the wrong way around. Maybe we could adjust the formula accordingly?